providers/implementations/signature/dsa_sig.c

```diff
@@ -331,8 +361,81 @@ static int dsa_sign(void *vpdsactx, unsigned char *sig, size_t *siglen,
     return 1;
 }
 
-static int dsa_verify(void *vpdsactx, const unsigned char *sig, size_t siglen,
-                      const unsigned char *tbs, size_t tbslen)
+static int dsa_signverify_message_update(void *vpdsactx,
+                                         const unsigned char *data,
+                                         size_t datalen)
+{
+    PROV_DSA_CTX *pdsactx = (PROV_DSA_CTX *)vpdsactx;
+
+    if (pdsactx == NULL)
+        return 0;
+
+    return EVP_DigestUpdate(pdsactx->mdctx, data, datalen);
+}
+
+static int dsa_sign_message_final(void *vpdsactx, unsigned char *sig,
+                                  size_t *siglen, size_t sigsize)
+{
+    PROV_DSA_CTX *pdsactx = (PROV_DSA_CTX *)vpdsactx;
+    unsigned char digest[EVP_MAX_MD_SIZE];
+    unsigned int dlen = 0;
+
+    if (!ossl_prov_is_running() || pdsactx == NULL)
+        return 0;
+    if (pdsactx->mdctx == NULL)
+        return 0;
+    /*
+     * If sig is NULL then we're just finding out the sig size. Other fields
+     * are ignored. Defer to dsa_sign.
+     */
+    if (sig != NULL) {
+        /*
+         * When this function is used through dsa_digest_sign_final(),
+         * there is the possibility that some externally provided digests
+         * exceed EVP_MAX_MD_SIZE. We should probably handle that
+         * somehow but that problem is much larger than just in DSA.
```

Comment:
Is it not possible to query the MD implementation for the digest size?  And if larger than the static `MAX`, dynamically allocate the right size?  Or error out?  Or anything better than running off  the end of the buffer???

Reply:
Sure.  But this is what was implemented before this PR, so may I suggest investigating and fixing in a separate PR?